### PR TITLE
Sort recipes

### PR DESCRIPTION
**Pull request by PatrikLundell**
#### Summary

SUMMARY: Interface "Sorts base camp and expansion construction recipes alphabetically"

#### Purpose of change

Allow base/expansion recipe creators to get similar recipes grouped together when presented to the player, rather than the current effectively random order by sorting the recipes alphabetically based on their short description (or "name").

In addition to that, a small bug fix snuck in (I'm not sure if it has been reported). If you send off a companion to create Tinder using the Canteen the companion is lost with no trace left on the billboard. You can pull the companion back by waiting 24 hours and do an emergency recall.
The crummy bug fix causes the expansion crafting order to appear in the list, and allows you to receive the tinder produced when bringing the companion back normally. However, due to how the implementation is made, there is no apparent way to distinguish between a companion sent off on a farm crafting mission from one sent off on a kitchen tinder making one, because the implementation used string matching against the exact same string literal in both cases, so both the farm and the "normal" expansion companion operations are invoked, and the one that isn't returning the companion returns an error message to the player. Still, it's better to get an error message than a lost companion, although the code as a whole should be revised.

#### Describe the solution

Sort the list of construction recipes based on the "name" before it's sent to the UI.

#### Describe alternatives you've considered

None

#### Testing

Verified that the construction recipes appear sorted and that invoking recipes result in the expected construction being constructed (rather than some other that might have been in that place before the list was sorted).

Verified that making Tinder from the kitchen actually works. Tried swapping the order in which the farm and directional crafting return orders were placed in the code with the same result, implying that farm crafting should behave the same as kitchen tinder, i.e. return the crafted goods, bug also give an error message.

#### Additional context

None on topic.

(Verified that the crafting logic is faulty by temporarily modify the recipe that allows the first cooking recipe group to also allow the first blacksmith recipe group and the blacksmith token. This resulted in both blacksmith and cooking recipes being displayed twice.
The error in the logic is that the list of all available recipes is generated, but then all recipes are presented once for each magic token present: there is nothing in the code trying to determine which recipes belong to each magic token lot, and nothing trying to tie the magic tokens to the tokens provided by recipe group "building_type".)

### CODE REVIEW

**Comment by PatrikLundell** [comment]
continuous-integration/appveyor/pr failure: "Could not locate cached archive: C:\Users\appveyor\AppData\Local\vcpkg\archives\52\52de75f386c8d5b7738208b01ab31999008592bf.zip"

Does not seem to be related to the PR.

File: src/faction_camp.cpp [hunk 1]
```diff
@@ -683,21 +683,36 @@ void basecamp::get_available_missions_by_dir( mission_data &mission_key, const p
...
+            order.push_back( i );
+        }
+
+        for( size_t i = 0; i + 1 < upgrades.size(); i++ ) {
+            for( size_t k = i + 1; k < upgrades.size(); k++ ) {
+                if( upgrades[order[k]].name.translated_lt( upgrades[order[i]].name ) ) {
+                    std::swap( order[i], order[k] );
+                }
+            }
+        }
```
Comment (BrettDong) [inline src/faction_camp.cpp:699]: I feel that this loop (and the similar one below) can be replaced by `std::sort` like
```suggestion
        std::sort( order.begin(), order.end(), [upgrades]( const size_t &p, const size_t &q )->bool {
            return upgrades[p].name.translated_lt( ( upgrades[q].name ) );
        } );
```

File: src/faction_camp.cpp [hunk 2]
```diff
@@ -683,21 +683,36 @@ void basecamp::get_available_missions_by_dir( mission_data &mission_key, const p
...
         // Generate upgrade missions for expansions
-        for( const basecamp_upgrade &upgrade : available_upgrades( dir ) ) {
-            const base_camps::miss_data &miss_info = base_camps::miss_info[ "_faction_upgrade_exp_" ];
-            comp_list npc_list = get_mission_workers( upgrade.bldg + "_faction_upgrade_exp_" +
+        std::vector<basecamp_upgrade> upgrades = available_upgrades( dir );
+        std::vector<size_t> order;
+
+        for( size_t i = 0; i < upgrades.size(); i++ ) {
+            order.push_back( i );
+        }
```
Comment (BrettDong) [inline src/faction_camp.cpp:691]: `order` vector can be initialised like
```suggestion
        std::vector<size_t> order(upgrades.size());
        std::iota(order.begin(), order.end(), 0);
```

**Comment by PatrikLundell** [comment]
The General build matrix / varied_builds (g++-7, 1, 1) (pull_request) error:
"
:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cata_test-tiles is a Catch v2.13.0 host application.
Run with -? for options

Randomness seeded to: 1609499710

-------------------------------------------------------------------------------
firebugs
-------------------------------------------------------------------------------
/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/tests/field_test.cpp:114
...............................................................................

/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/tests/field_test.cpp:111: FAILED:
  CHECK( !field_alive )
with expansion:
Error:   false
:
"
doesn't seem to have anything to do with the code in this PR.

**Comment by PatrikLundell** [comment]
General build matrix / varied_builds (g++-7, 1, 1) (pull_request) error:
"
:
All tests passed (3714190 assertions in 638 test cases)

Error: (continued from above) ERROR : src/map.cpp:5225 [void map::trap_set(const tripoint&, const trap_id&)] set trap funnel on top of terrain downspout which already has a builit-in trap
12:58:18.164 INFO : Ended test at Fri Jan  1 12:58:18 2021

12:58:18.164 INFO : The test took 686.016 seconds
12:58:18.164 INFO : Randomness seeded to: 1609505159
12:58:18.164 INFO : Treating result as failure due to error logged during tests.
Error: Process completed with exit code 1.
:
"
does not seem to have anything to do with the code in this PR.

**Comment by BrettDong** [comment]
I went through the code over again and don't understand why don't you just simply sort `upgrades`? Is there a reason behind I overlooked?

**Comment by PatrikLundell** [comment]
I think the only thing you've "overlooked" is history, i.e. the code looks the way it does because of how it evolved, not because it's the best way to write it.

Since the next step will be to get rid of "order", as it's used only to provide start and stop index functions for "sort", I'll ask about that here, rather than update the commit and get that as the next comment. The compiler complains about "0" for the first one but accepts the reasonable "upgrades.size() - 1" for the second one. It DOES accept "upgrades.size() - upgrades.size()" for the first one, but that's definitely ugly. What are reasonable candidates for these parameters?

**Comment by BrettDong** [comment]
Are you asking how to sort `upgrades`? I would write
```c++
        std::vector<basecamp_upgrade> upgrades = available_upgrades( dir );
        std::sort( upgrades.begin(), upgrades.end(), []( const basecamp_upgrade & p,
                   const basecamp_upgrade & q )->bool{return p.name.translated_lt( q.name );} );
```

**Comment by PatrikLundell** [comment]
Thanks, yes, that's it. I expected sort to use size_t parameters, but yes, it's using iterators. I'm not a C(++) programmer, and haven't really used anything newer than 2014...

**Comment by BrettDong** [comment]
Thanks! Now it looks much cleaner and simpler.

File: src/faction_camp.cpp [hunk 3]
```diff
@@ -675,16 +675,21 @@ void basecamp::get_available_missions_by_dir( mission_data &mission_key, const p
...
         }
         // Generate upgrade missions for expansions
-        for( const basecamp_upgrade &upgrade : available_upgrades( dir ) ) {
-            const base_camps::miss_data &miss_info = base_camps::miss_info[ "_faction_upgrade_exp_" ];
+        std::vector<basecamp_upgrade> upgrades = available_upgrades( dir );
+
+        std::sort( upgrades.begin(), upgrades.end(), []( const basecamp_upgrade & p,
+                   const basecamp_upgrade & q )->bool {return p.name.translated_lt( q.name ); } );
+
+        for( const basecamp_upgrade upgrade : upgrades ) {
```
Comment (BrettDong) [inline src/faction_camp.cpp:691]:
```suggestion
        for( const basecamp_upgrade &upgrade : upgrades ) {
```

File: src/faction_camp.cpp [hunk 4]
```diff
@@ -1200,8 +1215,13 @@ void basecamp::get_available_missions( mission_data &mission_key )
...
                                 base_camps::base_dir, entry, avail );
     }
-    for( const basecamp_upgrade &upgrade : available_upgrades( base_camps::base_dir ) ) {
-        const base_camps::miss_data &miss_info = base_camps::miss_info[ "_faction_upgrade_camp" ];
+    std::vector<basecamp_upgrade> upgrades = available_upgrades( base_camps::base_dir );
+
+    std::sort( upgrades.begin(), upgrades.end(), []( const basecamp_upgrade & p,
+               const basecamp_upgrade & q )->bool {return p.name.translated_lt( q.name ); } );
+
+    for( const basecamp_upgrade upgrade : upgrades ) {
```
Comment (BrettDong) [inline src/faction_camp.cpp:1223]:
```suggestion
    for( const basecamp_upgrade &upgrade : upgrades ) {
```

**Comment by BrettDong** [comment]
I was trying to compile the new code and compiler threw two warnings. After changing the above two lines of code it compiles.

**Comment by BrettDong** [comment]
I found that there are two entries of _"[B] (Finish) Crafting"_ after I dispatched one NPC to craft a tinder:
螢幕截圖 2021-01-05 上午11 52 15

Is this intentional?
```c++
    } else if( miss_id == miss_dir_id + " (Finish) Crafting" ) {
         const std::string craft_msg = _( "returns to you with something…" );
         mission_return( "_faction_camp_crafting_" + miss_dir_id, 1_minutes, true, craft_msg,
                         "construction", 2 );
         const std::string msg = _( "returns from your farm with something…" );
         mission_return( "_faction_exp_farm_crafting_" + miss_dir_id, 1_minutes, true, msg,
                         "construction", 2 );
    }
```

**Comment by PatrikLundell** [comment]
Yes and no...
It's a result of the somewhat crummy fix to the secondary issue in "Purpose of change".

A proper implementation would overhaul the complete string matching/stitching mess where UI strings are used as identifiers and replace it with tokens based on the recipe and the expansion, and possibly with a group, ideally structs, but if that causes backwards compatibility issues it would have to be strings.

I had intended to try to tackle that once my base camp PRs had all gone through, but after seeing that the core team is intent of ignoring those until a killer bot closes them I've lost interest.

**Review by BrettDong — APPROVED** [review]
I approve the sorting recipe part but I am not sure about the second part.

**Comment by PatrikLundell** [comment]
So I should revert it an have companions be sent to the void when tasked with making firewood from the kitchen? Sure, I can do that, and it won't hurt me, as I know it doesn't work.

Fixing the bug so it won't cause any side effects is a much larger project than the initial sorting + band aid, and is likely to have to be scrapped anyway when the code is reworked (eventually).

**Comment by PatrikLundell** [comment]
continuous-integration/travis-ci/pr — The Travis CI build failed
"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

cata_test.exe is a Catch v2.13.0 host application.

Run with -? for options

Randomness seeded to: 1609927563

-------------------------------------------------------------------------------

player::get_dodge

  each dodge after the first subtracts 2 points

-------------------------------------------------------------------------------

../tests/melee_dodge_hit_test.cpp:204

...............................................................................

../tests/melee_dodge_hit_test.cpp:207: FAILED:

  CHECK( dummy.get_dodge() == base_dodge - 2 )

with expansion:

  0.0f == 2.0f

1.426 s: map_bounds_checking

1.503 s: place_player_can_safely_move_multiple_submaps

"
does not seem to have anything to do with this PR.

**Comment by kevingranade** [comment]
> So I should revert it an have companions be sent to the void when tasked with making firewood from the kitchen?

I read this as "I'm signing off on the part I'm sure about, but not the other part.", not a *rejection* of the other part.

**Pull request merged**